### ISSUE TITLE
Limit Photos to a single folder + subtree

### ISSUE THREAD

**Issue by dermalikmann** [nextcloud/photos#141]
I think it would be really nice to create an option where the user can select a single folder (like `/Photos`) and the app only scans this folder and its subfolders.

I  have tons and tons of images all over my nextcloud, but only in `/Photos` are my, well, photos. And it's kind of a tedious task to create always a `.nomedia` file in all other directories.

**Comment by skjnldsv** [comment]
@jancborchardt ?

**Comment by jancborchardt** [comment]
Yes, @Mikescops and I briefly talked about this at the Contributor Week as well. What we could go for:
1. We look for a "Photos" or "Pictures" folder in the root. If it exists, we take that.
2. We should also account for translations, ideally following the [xdg-user-dirs](https://www.freedesktop.org/wiki/Software/xdg-user-dirs/) translations.
3. If no such folder exists, we fall back to root.
4. We show a settings area in the bottom left with:
  > Showing files from **Photos**
  > and a button: ( Change folder )

That way it’s an automatic setup process but can still be adjusted later.

**Comment by skjnldsv** [comment]
I would say it doesn't cover my needs. I have multiple folders with pictures that I want to be shown. Android is using all available photos, no idea for other galleries. Could be nice to check what users expect before pushing such drastic changes :)

Also, just having a setting to include/exclude some folders is probably the easiest/most universal way to do? :woman_shrugging:

**Comment by Mikescops** [comment]
@skjnldsv the point of @jancborchardt during our discussion was to have the most easy and transparent way to have the folder setups. I agree include/exclude would be the prefered way (as it is done on the sync for desktop client) but Jan thought choosing a single folder and having a per default on the Photos folder is the easiest way.

**Comment by skjnldsv** [comment]
Defaut is indeed always preferable, I'm not arguing this part.
But a default automated config needs to be relevant. I'm not sure this is relevant to use a default `Pictures` or `Photos` folder.
Even our Nextcloud client have the default autoupload set to `InstantUpload`, so we would not be compliant to our android setup :)

**Comment by Mikescops** [comment]
Let's take root folder but provide an option asap to restrict it to a given folder

**Comment by thutex** [comment]
same issue here that i am getting things like icons and unrelated artwork pictures in the photo app.
i suggest maybe copying the music app option to specify the path to your picture collection, and only if left empty use the root?

**Comment by saywebsolutions** [comment]
> same issue here that i am getting things like icons and unrelated artwork pictures in the photo app. i suggest maybe copying the music app option to specify the path to your picture collection, and only if left empty use the root?

Yes that would be great, or the other option mentioned above:

>Let's take root folder but provide an option asap to restrict it to a given folder

**Comment by bendschs** [comment]
does .noimage apply to a folder and all subfolders or just the folder itself?

**Comment by skjnldsv** [comment]
> does .noimage apply to a folder and all subfolders or just the folder itself?

folder and subfolders :)

**Comment by bendschs** [comment]
ok, in that case it simply does not work for me ... latest stable release of nextcloud 18

**Comment by strugee** [comment]
@bendschs please file a separate issue then. Make sure to search for duplicates beforehand.

**Comment by kicker334** [comment]
Can this be developed? the .noimage/.nomedia trick isn't working for any of my folders.

**Comment by jancborchardt** [comment]
@kicker334 sure thing – are you interested in contributing? :)

**Comment by kicker334** [comment]
@jancborchardt I honestly wouldn't know where to start, develop, or test this type of thing... :( I have to ask strangers on the internet for help.

**Comment by naezeroth** [comment]
I'd love to help out - I'm super new to this kind of thing but if I could get some starting tips I can definitely have a go!

**Comment by skjnldsv** [comment]
For anyone willing to help, please feel free to do so :)
It requires the addition of a setting section here
https://github.com/nextcloud/photos/blob/108105a984586db72fcc027cbbfeb87e737a35d5/src/Photos.vue#L27-L41
https://nextcloud-vue-components.netlify.app/#/Components/App%20containers?id=appnavigationsettings

Then a proper selection using our filePicker https://nextcloud.github.io/nextcloud-dialogs/modules/_filepicker_.html to specify which folder to use.

After that there will be various places to implement this filter:
https://github.com/nextcloud/photos/blob/108105a984586db72fcc027cbbfeb87e737a35d5/src/services/FileList.js#L53
https://github.com/nextcloud/photos/blob/108105a984586db72fcc027cbbfeb87e737a35d5/src/services/AlbumContent.js#L37 (see the AlbumController file, you should be able to specify the path)

This is definitely a complex task, but we'll be glad to assis anyone diving into this the best we can!

**Comment by ntimo** [comment]
This is a really important feature to me. Because otherwise the "Your Photos" view is completely broken for me since it displays all kind of images (like my 1 billion screenshots) that are not import and should not be shown there.

**Comment by muety** [comment]
I totally agree with @ntimo. I'd love to have Nextcloud photos be my primary photo storage, but due to this bug it's essentially unusable.

**Comment by saywebsolutions** [comment]
> I totally agree with @ntimo. I'd love to have Nextcloud photos be my primary photo storage, but due to this bug it's essentially unusable.

I concur; I fear that the developers aren't actually using the app to a sufficient extent to realize how poor the experience is. Sad, as I am really trying to like it

**Comment by silverdr** [comment]
Right - just add one more voice to the chorus. Keeping root as default but adding possibility to select what folders/subfolders should be used seems the best way to go.

**Comment by RebelCoderRU** [comment]
Hello. I looked through this a few times now, and I just can't figure out what is the solution to changing the Root photos folder for the Photos timeline? I am on the latest release 20. What needs to be done/configured to enable Photos timeline to be pointed to a custom photos folder?

**Comment by skjnldsv** [comment]
@RebelCoderRU 

This is an opened issue. This is not implemented

**Comment by StarSmasher44** [comment]
Any idea when this will be implemented at all? It's been almost a year now and in all honesty this isn't that big of a function that it would take so long to add something as simple as a (Customizable) list of scannable folders.

Without that, I have no use for the whole "Photos" app/functionality at all in NextCloud, simply because whatever I add gets buried within a whole bunch of random images..

If this really can't be done by the actual developers themselves, there is no shame but at least just tell us it isn't going to happen unless we build it ourselves.

Uninstalling the app until it actually works right, I suppose..

**Comment by strugee** [comment]
@StarSmasher44 if you read all the way through this issue you would see that #515 is open. You could try deploying that patch to production. YMMV though.

**Comment by Utopiah** [comment]
What about have a configuration item with the directory to start from and if it's not in the path in [scanCurrentFolder](https://github.com/nextcloud/photos/blob/master/lib/Controller/AlbumsController.php#L123) in AlbumsController then ignore the scan?

I'm posting this because even though a naive solution it seems that a lot of people would like to use Photos as a product with their NextCloud instance but have everything is blocking, which is a shame since most of the hard work is already done.

Suggestions done in other issues are surely better but also more complex (fine filtering) which seems not to get approved.

**Comment by skjnldsv** [comment]
> What about have a configuration item with the directory to start from and if it's not in the path in [scanCurrentFolder](https://github.com/nextcloud/photos/blob/master/lib/Controller/AlbumsController.php?rgh-link-date=2021-02-12T10%3A38%3A40Z#L123) in AlbumsController then ignore the scan?

This is not for the albums, this is for the Timeline, which uses a different endpoint
I'm not sure this feature is actually requested for the Albums section :thinking:

**Comment by Utopiah** [comment]
@skjnldsv then what's the indexing function for the Albums section? The idea would be the same, to filter out paths that are not in the configured one.

**Comment by skjnldsv** [comment]
This is the dav endpoint. Nextcloud's files indexing core.
Please check the other comments in https://github.com/nextcloud/photos/pull/515 so you can fully grasp the issues and why this hasn't been fixed yet :)

**Comment by b3nis** [comment]
I just wanted to "cheer on" this feature request. I don't have any development skills (I am just a useless product owner as profession). Happy to test tho :-)

**Comment by halfk1ng** [comment]
From a UX'er, this made my day

> useless product owner

**Comment by iluvatyr** [comment]
is photos app still developed? seems like this issue just doesnt get adressed.

**Comment by dermalikmann** [comment]
> is photos app still developed? seems like this issue just doesnt get adressed.

Have you actually read through the issue? There are several reasons why this hasn't been implemented yet.

**Comment by ParaplegicRacehorse** [comment]
[not a developer]  [but I can write help-docs]

I am also really excited to see this implemented.

Would it difficult to borrow code from the NC Music app, both for the scanning and the settings?

**Comment by mrtumnus** [comment]
IMO, this feature is no longer necessary due to the completion of the .nomedia/.noimage filter (#944).  Does exactly what I need (thought will require a bit more management to make sure any potential added folders get excluded if needed).

**Comment by inthreedee** [comment]
This is definitely still necessary. As stated in the original post:
> I have tons and tons of images all over my nextcloud, but only in /Photos are my, well, photos. And it's kind of a tedious task to create always a .nomedia file in all other directories.

An include list makes so much more sense than an exclude list, especially when the exclude list is handled by dot files in this way.

**Comment by brainfellout** [comment]
I agree 100% with what @inthreedee says above.

I provide Nextcloud services to various family members as well as myself, and none of us have any wish for the Photos app to show anything outside of the `~/Photos` directory (or the like). Using the `.nomedia` / `.noimage` filter is just not feasible for most casual users of a certain age / background.

Outside of family, I provide support to other computer users. I see they generally all have the same basic setup as each other, with all their photos in a single top level directory - usually the default photos folder generated by their OS. Images outside of this directory are generally not suitable for showing in a Photos app.

A nice easy include list that can be set once (or left to use a default value) seems to me like it would be the ideal way forward for 95% of users, with no future planning or maintenance required. As it stands, the `.nomedia` / `.noimage` filter won't be bothered with by many, meaning in turn, the Photos app remains next to useless.

**Comment by bendschs** [comment]
for me the  .nomedia/.noimage tag seems to be working for dekstop/browser (macos/brave) but not for the mobile app (ios).
just wanted to point that out.

( i am running  NC 24.0.6 and Apache PHP 8.0)

**Comment by ArtemSmaznov** [comment]
I hoped Photos 2.0 would fix this issue but seems there is no change. My photos are still swamped by the album arts from my Music directory.
A custom Photos directory is a must have for a photos app

**Comment by nssatlantis** [comment]
> I hoped Photos 2.0 would fix this issue but seems there is no change. My photos are still swamped by the album arts from my Music directory.
> A custom Photos directory is a must have for a photos app

You can use the .nomedia and .noimage files in said music folders so they will no longer show up.

**Comment by danieljkemp** [comment]
Needing to add files to folders explicitly is an incomplete solution when
people might have auto-created folders (like when your music software
imports a new album)

**Comment by ArtemSmaznov** [comment]
> > I hoped Photos 2.0 would fix this issue but seems there is no change. My photos are still swamped by the album arts from my Music directory.
> > A custom Photos directory is a must have for a photos app
> 
> You can use the .nomedia and .noimage files in said music folders so they will no longer show up.

I already tried that. As a test I created a new folder, added an empty .noimage file and uploaded a screenshot. It still showed in both All Media and Photos tabs.

Additionally, I'm not sure if this is supposed to work recursively (as in can I put 1 file in the root of my music dir or I need it in every album as I use beetz to organize my music and adding a file manually to each and every folder there seems like an unreasonable task)

**Comment by kitzler-walli** [comment]
Since the upgrade to nextcloud 25 the photos app ignores all .nomedia and .noimage files (it worked just fine in Version 24) and the new option "Photos directory" is ignored as well. No matter which folder I select I always see all images in all folders.

**Comment by bendschs** [comment]
> Since the upgrade to nextcloud 25 the photos app ignores all .nomedia and .noimage files (it worked just fine in Version 24) and the new option "Photos directory" is ignored as well. No matter which folder I select I always see all images in all folders.

can confirm, in NC25 it does not work for me even in the webbrowser any longer.

**Comment by ssc85** [comment]
I am also looking forward having a whitelist option to select root folders which are part of the photos/all media streams and not scanning all available files.

**Comment by ArtemSmaznov** [comment]
One thing that I haven't seen anyone mention yet. 
The android nextcloud app does have this feature - when inside the photos tab (called media) you can click the 3 dots on the top right and set the media folder and it updates the view to display only photos from that directory. Unfortunately it doesn't persist and you need to do it each time you go to media.
This makes me think that all we need is a similar option for web where you can specify which folder to display the photos from (would be too bad to have a list instead of just 1 path but that is just a cherry on top, at least for me personally)

**Comment by mrtumnus** [comment]
TIL the Nextcloud Android app has a Multimedia tab. A bit slow to load, but
promising nonetheless!

**Comment by snagglewest** [comment]
Not a whole lot to add but we're almost 3 years down the road and this is still a huge unresolved issue.  I'd love to get the "photos" app back but this version is simply useless for me as well as many other users.  Like most folks, I have tons and tons of images all over my various folders mapped into NC but only a few are actually photos.  Either living with thousands of images that aren't photos or trying to create a few hundred .nomedia files in all other directories is asinine.  If this app can't be fixed then NC needs to remove it as the standard because it's really not a "photos app", it's an image app...

**Comment by bendschs** [comment]
@meichthys thanx for the advice, had a quick look and i have to see, i actually don't see the need of an extra app (also considering all it's limitations of not running within the mobile nextcloud app on ios/android for instance) when up to 90% of it's functions are actually already covered within the fotos app. the straight forward thing to do from my point of view would be to manually pick the folders nextcloud should look for fotos. as @snagglewest pointed out, the app just becomes unusable for many users without that function. you will see all of your album covers, screenhots, icons and every other kind of image that actually just does not belong in the app.

**Comment by snagglewest** [comment]
@meichthys the memories app is a non-starter too.  My NC instance is just for family spread across the US.  The entire idea is to have a a few folders mapped from our NAS so everyone can see and enjoy the pics quick and easy with only 1 or 2 clicks.  If someone creates a new folder for say a vacation, and drops pics into it, everyone can see it and even add their own pics. Having to upload (duplicate), share, wiggle your nose etc is time consuming and really not user friendly, especially for older, less computer savvy folks in the family.   I guess it all comes down to who the target audience is.  For my target audience photos (or memories) is not it...

**Comment by tanuva** [comment]
Let me also add that the `.noimage` workaround has a significant drawback: It applies to all people a folder is shared with. That's not feasible for my NC instance. I would strongly prefer only to configure the photos folder for myself and let others make their own choice.

**Comment by silverdr** [comment]
> I guess it all comes down to who the target audience is

I rather disagree. If "Photos" is called "Photos" and not "All Possible Image Files You May Not Even Know You Have Somewhere On Your Filesystem(s)" then it should be easily configurable to show only "photos" one wishes to treat as such. Regardless of the target audience.

**Comment by T-bond** [comment]
Please stop spamming. It is a bugtracker, and not a forum. You are notifing 30+ people with your nonsense comments. Use reactions if you would like to move the thing forward, or better, submit a PR to fix the issue yourself. Maybe a maintainer could lock this thread?

We are half way there with the `.nomedia` solution.

**Comment by silverdr** [comment]
@T-bond do you believe your comment adds value for 30+ people and moves thing forward?

**Comment by nssatlantis** [comment]
> @T-bond do you believe your comment adds value for 30+ people and moves thing forward?

Then stop doing it jesus my dude

**Comment by saywebsolutions** [comment]
> Please stop spamming. It is a bugtracker, and not a forum. You are notifing 30+ people with your nonsense comments. Use reactions if you would like to move the thing forward, or better, submit a PR to fix the issue yourself. Maybe a maintainer could lock this thread?
> 
> We are half way there with the `.nomedia` solution.

Nobody here is spamming; there are just a lot of people frustrated/hopeful/interested in this issue as it's completely preventing the use of nextcloud photos in countless instances. Some of that interest is represented here in comments and discussion of the software (which is exactly what anyone who cares about nextcloud should want; the alternative is that nobody uses the software and nobody cares/nobody comments).

This feature (and Nextcloud Photos in general) is extra important as photos are a very important use case for Nextcloud so it's extra important that the core photos software works well and is very user friendly for non-technical users (so using hidden files is a non-starter; at best it's a kludge that technical users can implement). There really should be a folder configuration option with a sane default so most users won't need to touch it at all.

**Comment by jancborchardt** [comment]
Hi all – while this is a valid request, the discussion is getting very off-topic and heated as well, so I’ll lock it to contributors only. Please note this doesn’t mean we won’t work on it, so not to worry. :slightly_smiling_face: 

Thank you for understanding! :heart:

**Comment by szaimen** [comment]
This is possible since Photos 2.0 IIRC:
![image](https://user-images.githubusercontent.com/42591237/223508205-a2881dae-e8af-4e78-a9c9-5b9186ce7408.png)

**Comment by jancborchardt** [comment]
@AndyScherzinger @sorbaugh feature spec exists in my comment at https://github.com/nextcloud/photos/issues/141#issuecomment-577640415

**Comment by artonge** [comment]
@jancborchardt we already have a setting to configure upload directory for photos. I guess we should keep only one, right? Currently, the default value for this setting is the hard-coded string `'/Photos'`.

We can internationalize the default value with the way we usually do our translations.

![image](https://github.com/nextcloud/photos/assets/6653109/709e3887-4fc6-4094-8adb-3f54873757a5)

**Comment by szaimen** [comment]
I would personally vote for a second config option as people might want to set these independently...

**Comment by AndyScherzinger** [comment]
> We can internationalize the default value with the way we usually do our translations.

I wouldn't do that, on Android have have this -> AutoUpload folder is translated: So whenever you switch the language of the device, or have a new device or several with different languages it gets messy.

The only thing that comes to mind could be to have a way to display this fact as some meta info at the folder within the files app.

Just my 2 cents.

**Comment by artonge** [comment]
> I wouldn't do that, on Android have have this -> AutoUpload folder is translated: So whenever you switch the language of the device, or have a new device or several with different languages it gets messy.

Translated on init would be enough, no?

**Comment by AndyScherzinger** [comment]
> Translated on init would be enough, no?

Yes, that would be fine 👍

**Comment by artonge** [comment]
Should we also restric the search to the specified folder in other timeline views?
- "Favories", "On this day", "Shared with you"
- "Photos", I guess yes
- "Videos", it do not make sense, does it?

Same question for automated collections:
- "Places"
- "Peoples"

**Comment by artonge** [comment]
@jancborchardt or @nimishavijay

**Comment by nimishavijay** [comment]
As discussed, the folders from which media is shown can be configurable, and the folder to which photos are added is just one. Then it totally makes sense to restrict the timeline views, favourites, etc to only the selected folders :)

It could look like this in the settings :) 
- [ ] folders in the root folder do not have the path in the subline, otherwise the 2 parent folders are shown in the subline, so if `/Canon/Photography/2024/Portraits` is selected, subline is `Photography/2024`

![Untitled](https://github.com/nextcloud/photos/assets/52440189/58961478-ca58-41b9-bcbb-cc26e1a30859)

Nice-to-have little details: 
- [ ] show a different icon when the root folder is selected
- [ ] don't allow removing the folder if there is only one folder selected
![image](https://github.com/nextcloud/photos/assets/52440189/c0053df1-ef52-4320-a7bc-5e6902393b38)

ccing @jancborchardt in case you had something different in mind and also to double-check the wording :)

**Comment by jancborchardt** [comment]
@nimishavijay mockup looks great, we just need to make sure that the default for both Media folder and Upload folder is "Photos". :)

**Comment by sorbaugh** [comment]
@nimishavijay @jancborchardt I agree with removing the limitiation of having only 1 folder, and the nice-to-haves look neat. I'm splitting this into a separate subtask to get the current "finished" PR merged and make the individual steps and progress more transparent!

**Comment by dermalikmann** [comment]
🎉 🥳 
Thx everyone!

**Comment by jancborchardt** [comment]
@dermalikmann thanks for the really nice (and popular) suggestion! :) We hope you will enjoy the upcoming release!

**Comment by artonge** [comment]
@tobiasKaminsky I guess this needs to be taken into account by mobile clients. The setting is stored as a user config. Not sure whether you can easily access those from the clients ?

**Comment by melroy89** [comment]
OK.. interesting breaking change. Since now it gives me "No photos or videos in here" at all media & photos. While 1 user shared me folder with 10k images (basically all our personal/family photos), which I saw before. But not anymore ;(.

**Comment by ttrakker** [comment]
> I would say it doesn't cover my needs. I have multiple folders with pictures that I want to be shown. Android is using all available photos, no idea for other galleries. Could be nice to check what users expect before pushing such drastic changes :)
> 
> Also, just having a setting to include/exclude some folders is probably the easiest/most universal way to do? 🤷‍♀️

Since the world's been vaccinated with the genome shot, expect a lot of alternative systems to be sabotaged.

**Comment by ttrakker** [comment]
> @skjnldsv the point of @jancborchardt during our discussion was to have the most easy and transparent way to have the folder setups. I agree include/exclude would be the prefered way (as it is done on the sync for desktop client) but Jan thought choosing a single folder and having a per default on the Photos folder is the easiest way.

Easiest, not best. This degrades the system.
